src/java.base/share/classes/java/util/zip/ZipFile.java

```diff
@@ -1400,32 +1403,53 @@ private static class Key {
             File file;
             final boolean utf8;
 
-            public Key(File file, BasicFileAttributes attrs, ZipCoder zc) {
+            public Key(File file, BasicFileAttributes attrs, ZipCoder zc)
+                    throws IOException {
                 this.attrs = attrs;
-                this.file = file;
+                Object fk = attrs.fileKey();
+                if (fk != null) {
+                    // use file value, hashCode will use fileKey()
+                    this.file = file;
+                } else {
+                    this.file = getCanonicalFile(file);
+                }
                 this.utf8 = zc.isUTF8();
             }
 
+            @SuppressWarnings("removal")
+            private static File getCanonicalFile(File file) throws IOException {
+                try {
+                    return System.getSecurityManager() != null ?
+                        AccessController.doPrivileged((PrivilegedExceptionAction<File>) file::getCanonicalFile) :
```

Comment:
I think you'll have to cleanly separate the SM and no-SM case because of the exception handling, then you can use `PrivilegedExceptionAction<File> pae = file::getCanonicalFile`.